Mathlib/Data/Nat/GrayCode.lean

```diff
@@ -0,0 +1,163 @@
+/-
+Copyright (c) 2024 Daniel Weber. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Daniel Weber
+-/
+import Mathlib.Data.Nat.Bitwise
+import Mathlib.Data.Nat.Size
+import Mathlib.Data.Nat.Units
+
+/-!
+# Gray Code
+
+This file defines [the binary reflected gray code](https://en.wikipedia.org/wiki/Gray_code), both
+as a permutation of `ℕ`, in `Nat.gray_code`, and then, using that, as a permutation of `BitVec n`,
+in `Nat.partial_gray_code`. We also prove some theorems about them:
+
+* `Nat.gray_code_prop`: the xor of `Nat.gray_code n` and `Nat.gray_code (n+1)` is a power of 2
+for all `n`.
+* `gray_code_size`: `Nat.gray_code` preserves the binary length of integers.
+* `partial_gray_code_prop`: the xor of `n.partial_gray_code m` and `n.partial_gray_code (m+1)`
+is a power of 2 for all `n ≠ 0` and `m`.
+-/
+namespace Nat
+
+/-- The inverse of `f(n) = n ⊕ ⌊n/2⌋`, defined by `g(n) = n ⊕ g(⌊n/2⌋)`. -/
+def gray_code_inv : ℕ → ℕ
```

Comment:
defs (other than types) should be in lowerCamelCase rather than snake_case. So `grayCodeInv` here.